File: Drum/sticking-engine-ts/src/rule12-humanization.ts
/**
 * rule12-humanization.ts — RULE 12: CONTROLLED HUMAN VARIATION &
 * NON-ROBOTIC BEHAVIOR
 *
 * "Variation comes after validity and candidate generation; randomness
 * never repairs impossible behavior." This module runs strictly AFTER
 * Rule 8/29 has produced a physically valid committed sequence. It only
 * nudges microtiming and velocity within the style profile's configured
 * range, from a seeded generator, so runs are reproducible (Rule 26).
 */

import type { DrummerStyleProfile, HumanTimingContext } from './datamodel.js';
import { PythonRandom } from './core/mt19937.js';
import { sha256Hex } from './core/sha256.js';

/**
 * Every event gets its own reproducible sub-stream, derived from the
 * global seed plus the event id, so re-running with the same seed always
 * produces identical results regardless of solve order.
 *
 * The hash and the generator are both CPython's, reproduced exactly --
 * see `core/mt19937.ts` for why that is a contract and not a detail.
 */
function deterministicRng(seed: number, eventId: string): PythonRandom {
  const digest = sha256Hex(`${seed}:${eventId}`);
  return new PythonRandom(BigInt('0x' + digest.slice(0, 16)));
}

export const MAX_MICROTIMING_MS = 8.0;
export const MAX_VELOCITY_JITTER = 6;

/**
 * Bounded by `style.variationAmount` and the hard MAX_MICROTIMING_MS
 * ceiling -- never large enough to threaten Rule 6/13's physical
 * feasibility, only enough to avoid a quantized, robotic grid.
 */
export function humanizeTiming(
  eventId: string,
  scheduledTimeS: number,
  style: DrummerStyleProfile,
  seed: number,
  instrumentBiasMs = 0.0,
): HumanTimingContext {
  const rng = deterministicRng(seed, eventId + ':time');
  const boundMs = MAX_MICROTIMING_MS * style.variationAmount;
  const offsetMs = rng.uniform(-boundMs, boundMs) + instrumentBiasMs;
  return {
    eventId,
    scheduledTimeS,
    performedTimeS: scheduledTimeS + offsetMs / 1000.0,
    microtimingOffsetMs: offsetMs,
    instrumentBiasMs,
  };
}

/** The same, for dynamics. Clipped to the valid MIDI range. */
export function humanizeVelocity(
  eventId: string,
  baseVelocity: number,
  style: DrummerStyleProfile,
  seed: number,
): number {
  const rng = deterministicRng(seed, eventId + ':vel');
  const bound = MAX_VELOCITY_JITTER * style.variationAmount;
  const jitter = rng.uniform(-bound, bound);
  return Math.max(1, Math.min(127, pythonRound(baseVelocity + jitter)));
}

/**
 * Rule 12's "select among near-optimal valid alternatives": if the top
 * candidates are within epsilon of each other, pick among them
 * deterministically instead of always taking the greedy best -- which is
 * what keeps a repeated passage from looking robotically identical --
 * while never picking a worse-than-tied option.
 */
export function pickAmongNearTies<T>(
  candidates: readonly (readonly [T, number])[],
  seed: number,
  keyEventId: string,
  epsilon = 0.02,
): T | undefined {
  if (candidates.length === 0) return undefined;
  const bestScore = (candidates[0] as readonly [T, number])[1];
  const tied = candidates.filter((c) => bestScore - c[1] <= epsilon);
  if (tied.length === 1) return (tied[0] as readonly [T, number])[0];
  const rng = deterministicRng(seed, keyEventId + ':tie');
  return rng.choice(tied)[0];
}

/**
 * Python's `round()` — banker's rounding, half to EVEN.
 *
 * `round(99.5)` is 100 in JavaScript and 100 in Python, but
 * `round(98.5)` is 99 in JavaScript and 98 in Python. Velocities land on
 * exact halves often enough for this to matter, and a one-unit velocity
 * difference can cross a ghost/accent threshold in Rule 9.
 */
function pythonRound(x: number): number {
  const floor = Math.floor(x);
  const diff = x - floor;
  if (diff > 0.5) return floor + 1;
  if (diff < 0.5) return floor;
  return floor % 2 === 0 ? floor : floor + 1;
}

File: Drum/sticking-engine/sticking_engine/rule12_humanization.py
"""
rule12_humanization.py — RULE 12: CONTROLLED HUMAN VARIATION & NON-ROBOTIC
BEHAVIOR

Purpose: Add bounded, context-aware variation without destroying identity or
correctness.

Core Principle: "Variation comes after validity and candidate generation;
randomness never repairs impossible behavior." -> this module runs strictly
AFTER Rule 8/29 has already produced a physically valid committed sequence.
It only nudges microtiming/velocity within the style profile's configured
range, using a deterministic (seeded) RNG so runs are reproducible (Rule 26).
"""

from __future__ import annotations
import hashlib
import random
from dataclasses import dataclass
from typing import List

from .datamodel import DrummerStyleProfile, HumanTimingContext


def _deterministic_rng(seed: int, event_id: str) -> random.Random:
    """Every event gets its own reproducible sub-stream, derived from the
    global seed + event_id, so re-running with the same seed always produces
    identical results (Rule 26 Determinism requirement) regardless of solve
    order or parallelism."""
    h = hashlib.sha256(f"{seed}:{event_id}".encode()).hexdigest()
    return random.Random(int(h[:16], 16))


# Instrument/limb-independent nominal bias: humans tend to play slightly
# ahead of or behind the grid depending on articulation; kept tiny and bounded.
MAX_MICROTIMING_MS = 8.0
MAX_VELOCITY_JITTER = 6


def humanize_timing(event_id: str, scheduled_time_s: float, style: DrummerStyleProfile,
                     seed: int, instrument_bias_ms: float = 0.0) -> HumanTimingContext:
    """RULE 12 entry point for timing. Bounded by style.variation_amount and
    the hard MAX_MICROTIMING_MS ceiling — never large enough to threaten
    Rule 6/13's physical feasibility, only enough to avoid a quantized,
    robotic grid."""
    rng = _deterministic_rng(seed, event_id + ":time")
    bound_ms = MAX_MICROTIMING_MS * style.variation_amount
    offset_ms = rng.uniform(-bound_ms, bound_ms) + instrument_bias_ms
    performed = scheduled_time_s + (offset_ms / 1000.0)
    return HumanTimingContext(
        event_id=event_id,
        scheduled_time_s=scheduled_time_s,
        performed_time_s=performed,
        microtiming_offset_ms=offset_ms,
        instrument_bias_ms=instrument_bias_ms,
    )


def humanize_velocity(event_id: str, base_velocity: int, style: DrummerStyleProfile,
                       seed: int) -> int:
    """RULE 12 entry point for dynamics. Clipped to valid MIDI range."""
    rng = _deterministic_rng(seed, event_id + ":vel")
    bound = MAX_VELOCITY_JITTER * style.variation_amount
    jitter = rng.uniform(-bound, bound)
    return max(1, min(127, round(base_velocity + jitter)))


def pick_among_near_ties(candidates: List[tuple], seed: int, key_event_id: str,
                          epsilon: float = 0.02):
    """Rule 12's 'select among near-optimal valid alternatives' behavior.
    `candidates` is a list of (item, score) already sorted best-first.
    If the top candidates are within epsilon of each other, pick among them
    deterministically instead of always taking the single greedy best — this
    is what keeps repeated identical passages from looking robotically
    identical every single time, while never picking a worse-than-tied option.
    """
    if not candidates:
        return None
    best_score = candidates[0][1]
    tied = [c for c in candidates if best_score - c[1] <= epsilon]
    if len(tied) == 1:
        return tied[0][0]
    rng = _deterministic_rng(seed, key_event_id + ":tie")
    return rng.choice(tied)[0]
